pkg/segment/query/iqr/intermediateQueryResult.go

```diff
@@ -0,0 +1,372 @@
+		}
+
+		vTable := iqr.rrcs[0].VirtualTableName
+		colToValues, err := record.ReadAllColsForRRCs(segKey, vTable, rrcs, iqr.qid)
+		if err != nil {
+			log.Errorf("IQR.readAllColumnsWithRRCs: error reading all columns for segKey %v; err=%v",
+				segKey, err)
+			return nil
+		}
+
+		return colToValues
+	}
+
+	results := toputils.BatchProcessToMap(iqr.rrcs, getBatchKey, batchKeyLess, batchOperation)
+
+	for _, values := range results {
+		if len(values) != len(iqr.rrcs) {
+			// This will happen if we got an error in the batch operation.
+			return nil, toputils.TeeErrorf("IQR.readAllColumnsWithRRCs: expected %v results, got %v",
+				len(iqr.rrcs), len(values))
+		}
+	}
+
+	return results, nil
+}
+
+func (iqr *IQR) readColumnWithRRCs(cname string) ([]utils.CValueEnclosure, error) {
+	// Prepare to call BatchProcess().
+	getBatchKey := func(rrc *utils.RecordResultContainer) uint16 {
+		return rrc.SegKeyInfo.SegKeyEnc
+	}
+	batchKeyLess := toputils.NewUnsetOption[func(uint16, uint16) bool]()
+	batchOperation := func(rrcs []*utils.RecordResultContainer) []utils.CValueEnclosure {
+		if len(rrcs) == 0 {
+			return nil
+		}
+
+		segKey, ok := iqr.encodingToSegKey[rrcs[0].SegKeyInfo.SegKeyEnc]
+		if !ok {
+			log.Errorf("IQR.readColumnWithRRCs: unknown encoding %v", rrcs[0].SegKeyInfo.SegKeyEnc)
+			return nil
+		}
+
+		values, err := record.ReadColForRRCs(segKey, rrcs, cname, iqr.qid)
+		if err != nil {
+			log.Errorf("IQR.readColumnWithRRCs: error reading column %s: %v", cname, err)
+			return nil
+		}
+
+		return values
+	}
+
+	results := toputils.BatchProcess(iqr.rrcs, getBatchKey, batchKeyLess, batchOperation)
+
+	if len(results) != len(iqr.rrcs) {
+		// This will happen if we got an error in the batch operation.
+		return nil, toputils.TeeErrorf("IQR.readColumnWithRRCs: expected %v results, got %v",
+			len(iqr.rrcs), len(results))
+	}
+
+	// TODO: should we have an option to disable this caching?
+	iqr.knownValues[cname] = results
+
+	return results, nil
+}
+
+func (iqr *IQR) AsResult() (*pipesearch.PipeSearchResponseOuter, error) {
+	if err := iqr.validate(); err != nil {
+		log.Errorf("IQR.AsResult: invalid state: %v", err)
+		return nil, err
+	}
+
+	var records map[string][]utils.CValueEnclosure
+	var err error
+	switch iqr.mode {
+	case notSet:
+		// There's no data.
+		return nil, nil
+	case withRRCs:
+		records, err = iqr.readAllColumnsWithRRCs()
+		if err != nil {
+			log.Errorf("IQR.AsResult: error reading all columns: %v", err)
+			return nil, err
+		}
+	case withoutRRCs:
+		records = iqr.knownValues
+	default:
+		return nil, fmt.Errorf("IQR.AsResult: unexpected mode %v", iqr.mode)
+	}
+
+	cValRecords := toputils.TransposeMapOfSlices(records)
+	recordsAsAny := make([]map[string]interface{}, len(cValRecords))
+	for i, record := range cValRecords {
+		recordsAsAny[i] = make(map[string]interface{})
+		for key, value := range record {
+			recordsAsAny[i][key] = value.CVal
+		}
+	}
+
+	response := &pipesearch.PipeSearchResponseOuter{
```

Comment:
do we need to add a TODO for other response handler types like es bulk response handler , pipesearchWS response ?

Reply:
For websocket searches I think we do need another one, so I added a TODO. I'm not sure what you meant for ES bulk; we'll only want to convert the IQR to a result for queries, and I thought we only had API queries and WS queries

Reply:
so we do have the es query response handler, this is mostly used for the kibana and ES api compatibility. But we have de-priroritized that. For now just putting a todo comment for es resp handler is good enough